### ISSUE TITLE
Macos build torch in debug mode can not gdb in C++,But I can debug c and cpp by gdb

### ISSUE THREAD

**Issue by HongLouyemeng** [pytorch/pytorch#115476]
### 🐛 Describe the bug

```python
import torch
a = torch.Tensor([2,3])
print(a)
```
gdb --args python3 test.py
 sudo DEBUG=1  MACOSX_DEPLOYMENT_TARGET=14.1  CC=clang CXX=clang++  USE_CUDA=0 USE_DISTRIBUTED=1 python3 setup.py build develop && python -c "import torch"
![image](https://github.com/pytorch/pytorch/assets/48181529/9cdde4f8-7a1b-4935-a805-7c849db87594)
![image](https://github.com/pytorch/pytorch/assets/48181529/38174c61-e8a4-4dea-b896-129ef28cfaac)

### Versions

Collecting environment information...
PyTorch version: 2.2.0a0+git317486e
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.1 (x86_64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.0.40.1)
CMake version: version 3.26.4
Libc version: N/A

Python version: 3.8.18 (default, Sep 11 2023, 08:17:33)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.4
[pip3] numpydoc==1.5.0
[pip3] torch==2.2.0a0+git317486e
[conda] blas                      1.0                         mkl    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl                       2021.4.0           hecd8cb5_637    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl-include               2023.1.0         hecd8cb5_43560    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl-service               2.4.0            py38h9ed2024_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_fft                   1.3.1            py38h4ab4a9b_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_random                1.2.2            py38hb2f4e1b_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] numpy                     1.24.4                   pypi_0    pypi
[conda] numpydoc                  1.5.0            py38hecd8cb5_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] torch                     2.2.0a0+git317486e           dev_0    <develop>

cc @seemethere @malfet @osalpekar @atalman @albanD

**Comment by malfet** [comment]
Sorry, but can you provide a bit more details? Also, I'm not sure about status of gdb for MacOS, but I'm using lldb on an almost a daily basis to debug on MacOS and it works as expected

**Comment by HongLouyemeng** [comment]
It seems that it cannot capture the C++source code of Torch. Thank you for your experience, I will try the effect of LLDB
<img width="464" alt="image" src="https://github.com/pytorch/pytorch/assets/48181529/66e1a726-6117-4a95-a99d-842e5c14bb4d">

**Comment by HongLouyemeng** [comment]
Hi,lldb works in my mac, gdb is not work.thank you.
<img width="521" alt="image" src="https://github.com/pytorch/pytorch/assets/48181529/516ccabf-152f-4be5-b524-d99d0414f2ca">
<img width="501" alt="image" src="https://github.com/pytorch/pytorch/assets/48181529/ceda325c-f2a6-4223-b360-4a4648a71da1">

**Comment by HongLouyemeng** [comment]
But I can't debug the same code in torch.distributed, Did lldb cannot debug distributed application?QAQ
```python
import os
import time
import numpy as np
import torch
import torch.multiprocessing as mp
import torch.distributed as dist
from torch.distributed._tensor import DTensor, DeviceMesh, Shard, Replicate, distribute_tensor,zeros

def run(rank, size):

  a = torch.tensor([[0, 2.], [3, 0]])
  a.neg()


def init_process(rank_id, size, fn, backend='gloo'):
    """ Initialize the distributed environment. """
    os.environ['MASTER_ADDR'] = '127.0.0.1'
    os.environ['MASTER_PORT'] = '12347'
    os.environ[
        "TORCH_DISTRIBUTED_DEBUG"
    ] = "DETAIL" 
    dist.init_process_group(backend, rank=rank_id, world_size=size)
    fn(rank_id, size)

if __name__ == "__main__":
    big_tensor = torch.arange(0,16).reshape(4,4)
    size = 1
    processes = []
    mp.set_start_method("spawn")
    for rank in range(size):
        p = mp.Process(target=init_process, args=(rank, size, run))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()
```

**Comment by HongLouyemeng** [comment]
when I  use lldb attach subprocess, Is stopedQAQ
<img width="576" alt="image" src="https://github.com/pytorch/pytorch/assets/48181529/c16b28bb-be25-49c3-8351-fcd6a0cd5d90">

**Comment by malfet** [comment]
Thank you for providing a details. It feels like the original issue  you are describing here would be a great subject of https://discuss.pytorch.org/ but it does not look like a PyTorch issue, but rather GDBs inability to parse some of the data formats available in MacOS Sonoma.

There also feels another ask here for better documentation on how to do distributed debugging, if you think such tutorial/documentation update would be helpful, please do not hesitate to file a feature request.

I don't know, what you are trying to debug, but I wonder if doing the same exercise inside say [python](https://hub.docker.com/_/python) docker container would unblock you.

Please do not hesitate to open a new issue if you believe PyTorch does not function as intended.
